private/buf/bufgen/features.go

```diff
@@ -15,93 +15,189 @@
+			featureMask := uint64(feature)
 			if supported&featureMask != featureMask {
 				// doh! Supported features don't include this one
-				failed[feature] = files
+				failed.featureToFilename[feature] = files
 				failedFeatures = append(failedFeatures, feature)
 			}
 		}
-		if len(failed) > 0 {
-			// TODO: plugin config to turn this into an error
-			_, _ = fmt.Fprintf(container.Stderr(), "Warning: plugin %q does not support required features.\n",
-				configs[responseIndex].Name())
+		if supported&uint64(pluginpb.CodeGeneratorResponse_FEATURE_SUPPORTS_EDITIONS) != 0 && len(required.editionToFilename) > 0 {
+			// Plugin supports editions, and files include editions. So make sure
+			// the plugin supports precisely the right editions.
+			requiredEditions := make([]descriptorpb.Edition, 0, len(required.editionToFilename))
+			for edition := range required.editionToFilename {
+				requiredEditions = append(requiredEditions, edition)
+			}
+			sort.Slice(requiredEditions, func(i, j int) bool {
+				return requiredEditions[i] < requiredEditions[j]
+			})
+			for _, requiredEdition := range requiredEditions {
+				if int32(requiredEdition) < response.GetMinimumEdition() ||
+					int32(requiredEdition) > response.GetMaximumEdition() {
+					failed.editionToFilename[requiredEdition] = required.editionToFilename[requiredEdition]
+					failedEditions = append(failedEditions, requiredEdition)
+				}
+			}
+		}
+
+		pluginName := configs[responseIndex].Name()
+		if len(failedFeatures) > 0 {
+			_, _ = fmt.Fprintf(
+				&errorDetails,
+				"Plugin %q does not support required feature(s).\n",
+				pluginName)
 			sort.Slice(failedFeatures, func(i, j int) bool {
-				return failedFeatures[i].Number() < failedFeatures[j].Number()
+				return failedFeatures[i] < failedFeatures[j]
 			})
 			for _, feature := range failedFeatures {
-				files := failed[feature]
-				_, _ = fmt.Fprintf(container.Stderr(), "  Feature %q is required by %d file(s):\n",
+				files := failed.featureToFilename[feature]
+				_, _ = fmt.Fprintf(
+					&errorDetails,
+					"  Feature %q is required by %d file(s):\n",
 					featureName(feature), len(files))
-				_, _ = fmt.Fprintf(container.Stderr(), "    %s\n", strings.Join(files, ","))
+				_, _ = fmt.Fprintf(
+					&errorDetails,
+					"    %s\n",
+					strings.Join(files, ","))
+			}
+		}
+
+		if len(failedEditions) > 0 {
+			_, _ = fmt.Fprintf(
+				&errorDetails,
+				"Plugin %q does not support required edition(s).\n",
+				pluginName)
+			sort.Slice(failedEditions, func(i, j int) bool {
+				return failedEditions[i] < failedEditions[j]
+			})
+			for _, edition := range failedEditions {
+				files := failed.editionToFilename[edition]
+				_, _ = fmt.Fprintf(
+					&errorDetails,
+					"  Edition %q is required by %d file(s):\n",
+					editionName(edition), len(files))
+				_, _ = fmt.Fprintf(
+					&errorDetails,
+					"    %s\n",
+					strings.Join(files, ","))
 			}
 		}
 	}
+	if errorDetails.Len() > 0 {
```

Comment:
Just use []error and multierr.Join, I haven't seen fmt.Printf to a buffer for error messages, and would argue it's strictly less unwindable than a multierror

Reply:
So a separate error for every instance? Would you expect the output then to change from this:
```
Plugin "test" does not support required feature(s).
  Feature "proto3 optional" is required by 3 file(s):
    foo.proto, bar.proto, baz.proto
  Feature "supports editions" is required by 1 file(s):
    editions.proto
```
to this?
```
plugin "test" does not support feature "proto3 optional" which is required by "foo.proto"
plugin "test" does not support feature "proto3 optional" which is required by "bar.proto"
plugin "test" does not support feature "proto3 optional" which is required by "baz.proto"
plugin "test" does not support feature "supports editions" which is required by "editions.proto"
```

Reply:
Done. Output changed to above, hopefully that's what you were looking for.